File: frontend/src/api/coffeeTypes.ts
export type KubeObjectMeta = {
  uid?: string
  resourceVersion?: string
  name?: string
  namespace?: string
  generation?: number
  labels?: Record<string, string>
}

export type SecretKeyRef = {
  name?: string
  key?: string
}

export type CoffeeProductSpec = {
  sku: string
  name: string
  priceCents: number
  description?: string
  enabled: boolean
}

export type CoffeeVoucherSpec = {
  code: string
  enabled: boolean
  discountType: string
  discountValue: number
  maximumUsage: number
  appliesToProducts: string[]
  displayMessage?: string
}

export type CoffeeMailSpec = {
  provider?: string
  apiKeySecretRef?: SecretKeyRef
  orderConfirmationTemplate?: string
  fromAddress?: string
}

export type CoffeePaymentsSpec = {
  provider?: string
  apiKeySecretRef?: SecretKeyRef
  mode?: string
  zeroAmountCheckoutAllowed?: boolean
}

export type CoffeeConfigSpec = {
  shopName?: string
  bannerText?: string
  qrUrl?: string
  currency?: string
  products: CoffeeProductSpec[]
  vouchers: CoffeeVoucherSpec[]
  mail?: CoffeeMailSpec
  payments?: CoffeePaymentsSpec
}

export type CoffeeConfig = {
  apiVersion?: string
  kind?: string
  metadata?: KubeObjectMeta
  spec: CoffeeConfigSpec
}

export type StorefrontProduct = {
  sku: string
  name: string
  description?: string
  enabled: boolean
  basePriceCents: number
  displayPriceCents: number
  voucherState: string
}

export type StorefrontResponse = {
  shop: {
    name: string
    bannerText: string
    currency: string
  }
  voucher: {
    code: string
    presentInUrl: boolean
    displayMessage: string
    state: string
  }
  products: StorefrontProduct[]
}

export type CoffeeOrderItemRequest = {
  sku: string
  quantity: number
}

export type CoffeeOrderRequest = {
  voucherCode?: string
  source?: Record<string, string>
  items: CoffeeOrderItemRequest[]
}

export type CoffeeOrderFailure = {
  code: string
  message: string
}

export type CoffeeOrderResponse = {
  orderId: string
  status: 'placed' | 'rejected'
  currency: string
  totalPriceCents: number
  failure?: CoffeeOrderFailure
}

/** One line of an order, as the feed reports it. */
export type CoffeeOrderLine = {
  sku: string
  name: string
  quantity: number
  unitPriceCents: number
  lineTotalCents: number
  voucherApplied: boolean
}

/**
 * One submission on the live order feed.
 *
 * Note what this is NOT: there is no apiVersion, no kind, no metadata, no
 * resourceVersion. An order is an event in the Voter process, not a Kubernetes
 * object, and the shape says so. `who` is a display name and the only thing the
 * feed knows about a person.
 */
export type CoffeeOrderRecord = {
  /** This process's own counter, and the only key a rejected order has. */
  seq: number
  orderId?: string
  submittedAt: string
  who?: string
  voucherCode?: string
  items?: CoffeeOrderLine[]
  currency?: string
  totalPriceCents: number
  status: 'placed' | 'rejected'
  failureCode?: string
  failureMessage?: string
}

/** `scope: "process"` is a warning: these are one replica's orders, since its
 *  last restart. Nothing here survives a pod restart, and nothing is in Git. */
export type CoffeeOrdersSnapshot = {
  orders: CoffeeOrderRecord[]
  scope: string
  capacity: number
}

File: voter/coffee_types.go
package main

type kubeObjectMeta struct {
	Name            string            `json:"name,omitempty"`
	Namespace       string            `json:"namespace,omitempty"`
	Generation      int64             `json:"generation,omitempty"`
	ResourceVersion string            `json:"resourceVersion,omitempty"`
	Labels          map[string]string `json:"labels,omitempty"`
}

type secretKeyRef struct {
	Name string `json:"name,omitempty"`
	Key  string `json:"key,omitempty"`
}

type coffeeProductSpec struct {
	SKU         string `json:"sku,omitempty"`
	Name        string `json:"name,omitempty"`
	PriceCents  int    `json:"priceCents,omitempty"`
	Description string `json:"description,omitempty"`
	Enabled     bool   `json:"enabled,omitempty"`
}

type coffeeVoucherSpec struct {
	Code              string   `json:"code,omitempty"`
	Enabled           bool     `json:"enabled,omitempty"`
	DiscountType      string   `json:"discountType,omitempty"`
	DiscountValue     int      `json:"discountValue,omitempty"`
	MaximumUsage      int      `json:"maximumUsage,omitempty"`
	AppliesToProducts []string `json:"appliesToProducts,omitempty"`
	DisplayMessage    string   `json:"displayMessage,omitempty"`
}

type coffeeMailSpec struct {
	Provider                  string        `json:"provider,omitempty"`
	APIKeySecretRef           *secretKeyRef `json:"apiKeySecretRef,omitempty"`
	OrderConfirmationTemplate string        `json:"orderConfirmationTemplate,omitempty"`
	FromAddress               string        `json:"fromAddress,omitempty"`
}

type coffeePaymentsSpec struct {
	Provider                  string        `json:"provider,omitempty"`
	APIKeySecretRef           *secretKeyRef `json:"apiKeySecretRef,omitempty"`
	Mode                      string        `json:"mode,omitempty"`
	ZeroAmountCheckoutAllowed bool          `json:"zeroAmountCheckoutAllowed,omitempty"`
}

type coffeeConfigSpec struct {
	ShopName   string              `json:"shopName,omitempty"`
	BannerText string              `json:"bannerText,omitempty"`
	QRURL      string              `json:"qrUrl,omitempty"`
	Currency   string              `json:"currency,omitempty"`
	Products   []coffeeProductSpec `json:"products,omitempty"`
	Vouchers   []coffeeVoucherSpec `json:"vouchers,omitempty"`
	Mail       coffeeMailSpec      `json:"mail,omitempty"`
	Payments   coffeePaymentsSpec  `json:"payments,omitempty"`
}

type coffeeConfig struct {
	APIVersion string           `json:"apiVersion,omitempty"`
	Kind       string           `json:"kind,omitempty"`
	Metadata   kubeObjectMeta   `json:"metadata,omitempty"`
	Spec       coffeeConfigSpec `json:"spec,omitempty"`
}

type storefrontResponse struct {
	Shop     storefrontShop      `json:"shop"`
	Voucher  storefrontVoucher   `json:"voucher"`
	Products []storefrontProduct `json:"products"`
}

type storefrontShop struct {
	Name       string `json:"name"`
	BannerText string `json:"bannerText"`
	Currency   string `json:"currency"`
}

type storefrontVoucher struct {
	Code           string `json:"code"`
	PresentInURL   bool   `json:"presentInUrl"`
	DisplayMessage string `json:"displayMessage"`
	State          string `json:"state"`
}

type storefrontProduct struct {
	SKU               string `json:"sku"`
	Name              string `json:"name"`
	Description       string `json:"description"`
	Enabled           bool   `json:"enabled"`
	BasePriceCents    int    `json:"basePriceCents"`
	DisplayPriceCents int    `json:"displayPriceCents"`
	VoucherState      string `json:"voucherState"`
}

type coffeeOrderRequest struct {
	VoucherCode string                   `json:"voucherCode"`
	Source      map[string]string        `json:"source"`
	Items       []coffeeOrderItemRequest `json:"items"`
}

type coffeeOrderItemRequest struct {
	SKU      string `json:"sku"`
	Quantity int    `json:"quantity"`
}

type coffeeOrderLine struct {
	SKU            string `json:"sku"`
	Name           string `json:"name"`
	Quantity       int    `json:"quantity"`
	UnitPriceCents int    `json:"unitPriceCents"`
	LineTotalCents int    `json:"lineTotalCents"`
	VoucherApplied bool   `json:"voucherApplied"`
}

type coffeeOrderFailure struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}
